Firestore/core/src/firebase/firestore/immutable/sorted_set.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_SORTED_SET_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_SORTED_SET_H_
+
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/sorted_map.h"
+#include "Firestore/core/src/firebase/firestore/immutable/sorted_map_base.h"
+#include "Firestore/core/src/firebase/firestore/util/comparison.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+
+namespace impl {
+
+// An empty value to associate with keys in the underlying map.
+struct Empty {
+  friend bool operator==(Empty /* left */, Empty /* right */) {
+    return true;
+  }
+};
+
+}  // namespace impl
+
+template <typename K,
+          typename V = impl::Empty,
+          typename C = util::Comparator<K>,
+          typename M = SortedMap<K, V, C>>
+class SortedSet {
+ public:
+  using size_type = typename M::size_type;
+  using value_type = K;
```

Comment:
`value_type` is the type of the elements of a collection so I prefer it, also because it's `values_from` and `values_in` not `keys_from` and `keys_in`.

The type parameter must be `K` only because we also need `V` to denote the mapped type in the underlying map.